components/brave_wallet/browser/cardano/cardano_provider_impl.cc

```diff
@@ -91,371 +65,43 @@ void CardanoProviderImpl::Enable(EnableCallback callback) {
-          ->SignCip30MessageByCardanoKeyring(account_id, key_id, message);
-
-  if (!sig_data) {
-    std::move(callback).Run(
-        std::nullopt,
-        mojom::CardanoProviderErrorBundle::New(
-            kAPIErrorInternalError, WalletInternalErrorMessage(), nullptr));
-    return;
-  }
-
-  std::move(callback).Run(std::move(*sig_data), nullptr);
-}
-
-void CardanoProviderImpl::GetCollateral(const std::string& amount,
-                                        GetCollateralCallback callback) {
-  auto account_id = GetAllowedSelectedAccount();
-  if (!account_id) {
-    std::move(callback).Run(
-        std::nullopt,
-        mojom::CardanoProviderErrorBundle::New(
-            kAPIErrorRefused, kAccountNotConnectedError, nullptr));
-    return;
-  }
-
-  delegate_->WalletInteractionDetected();
-
-  NOTIMPLEMENTED_LOG_ONCE();
-  std::move(callback).Run(
-      std::nullopt, mojom::CardanoProviderErrorBundle::New(
-                        kAPIErrorInternalError, kNotImplemented, nullptr));
-}
-
-void CardanoProviderImpl::RequestCardanoPermissions(EnableCallback callback,
-                                                    const url::Origin& origin) {
+void CardanoProviderImpl::RequestCardanoPermissions(
+    mojo::PendingReceiver<mojom::CardanoApi> cardano_api,
+    EnableCallback callback,
+    const url::Origin& origin) {
   if (delegate_->IsPermissionDenied(mojom::CoinType::ADA)) {
-    OnRequestCardanoPermissions(std::move(callback), origin,
-                                mojom::RequestPermissionsError::kNone,
+    OnRequestCardanoPermissions(std::move(cardano_api), std::move(callback),
+                                origin, mojom::RequestPermissionsError::kNone,
                                 std::vector<std::string>());
     return;
   }
 
-  // TODO(cypt4): Support multiple Cardano accounts
-  auto cardano_account_id =
-      FindFirstCardanoHDAccount(brave_wallet_service_->keyring_service());
+  auto cardano_account_ids = GetCardanoAccountPermissionIdentifiers(
+      brave_wallet_service_->keyring_service());
 
-  if (!cardano_account_id) {
+  if (cardano_account_ids.empty()) {
     if (!wallet_onboarding_shown_) {
       delegate_->ShowWalletOnboarding();
       wallet_onboarding_shown_ = true;
     }
-    OnRequestCardanoPermissions(std::move(callback), origin,
-                                mojom::RequestPermissionsError::kInternal,
-                                std::nullopt);
+    OnRequestCardanoPermissions(
+        std::move(cardano_api), std::move(callback), origin,
+        mojom::RequestPermissionsError::kInternal, std::nullopt);
     return;
   }
 
   if (brave_wallet_service_->keyring_service()->IsLockedSync()) {
     if (pending_request_cardano_permissions_callback_) {
       OnRequestCardanoPermissions(
-          std::move(callback), origin,
+          std::move(cardano_api), std::move(callback), origin,
           mojom::RequestPermissionsError::kRequestInProgress, std::nullopt);
       return;
```

Comment:
We don't need to call OnRequestCardanoPermissions here and in all other error cases. Let's std::move(callback).Run() with error inplace